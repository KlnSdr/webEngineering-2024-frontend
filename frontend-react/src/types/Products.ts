interface NeededProduct {
  id: number; // needed because react decided to be difficult today!
  productName: string;
  amount: number;
}

interface Product {
    name: string;
    unit: string;
}

export type { NeededProduct, Product };
