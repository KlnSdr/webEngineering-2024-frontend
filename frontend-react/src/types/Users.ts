interface User {
    userId: number;
    userName: string;
}

export type { User };
