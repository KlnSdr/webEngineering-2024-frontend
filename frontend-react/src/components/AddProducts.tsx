import React, { useState } from "react";
import AddProductLine from "./AddProductLine";
import Button from "react-bootstrap/Button";
import { NeededProduct, Product } from "../types/Products";

interface AddProductsProps {
  onChange: (products: NeededProduct[]) => void;
  availableProducts: Product[];
}

const AddProducts: React.FC<AddProductsProps> = ({
  onChange,
  availableProducts,
}) => {
  const [neededProducts, setNeededProducts] = useState<NeededProduct[]>([]);
  const emptyNeededProduct: NeededProduct = {
    id: -1,
    productName: "",
    amount: 0,
  };

  const updateNeededProduct = (
    index: number,
    productName: string,
    amount: number
  ) => {
    const neededProductsCopy: NeededProduct[] = [...neededProducts];
    neededProductsCopy[index] = {
      ...neededProductsCopy[index],
      productName: productName,
      amount: amount,
    };
    setNeededProducts(neededProductsCopy);
    onChange(neededProductsCopy);
  };

  const removeNeededProductAt = (index: number) => {
    const filtered = neededProducts.filter((_, i: number) => i !== index);
    setNeededProducts(filtered);
    onChange(filtered);
  };

  return (
    <div>
      {neededProducts.map((product: NeededProduct, index: number) => {
        return (
          <AddProductLine
            key={product.id}
            products={availableProducts.map((prod: Product) => prod.name)}
            initialValue={product.productName}
            onChange={(prod: string, amount: number) => {
              updateNeededProduct(index, prod, amount);
            }}
            onRemove={() => {
              removeNeededProductAt(index);
            }}
            getUnitOf={(product: string) => {
              return (
                availableProducts.find((prod: Product) => prod.name === product)
                  ?.unit || ""
              );
            }}
          />
        );
      })}

      <Button
        onClick={() => {
          const neededProductsNew: NeededProduct[] = [
            ...neededProducts,
            { ...emptyNeededProduct, id: Date.now() },
          ];
          setNeededProducts(neededProductsNew);
          onChange(neededProductsNew);
        }}
        className="bi bi-plus"
      ></Button>
    </div>
  );
};

export default AddProducts;
