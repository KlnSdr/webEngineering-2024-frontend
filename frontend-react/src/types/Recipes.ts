import { NeededProduct } from "./Products";

interface CreateRecipe {
  title: string;
  image: File | null;
  description: string;
  products: NeededProduct[];
}

export type { CreateRecipe };
