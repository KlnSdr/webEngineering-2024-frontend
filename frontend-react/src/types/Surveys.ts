import {Recipe} from "./Recipes";

interface Survey{
    id: number;
    title: string;
    participants: string[];
    creator: string;
    recipeVote: {[key: string]: number};
    options: string[];
    creationDate: Date;
}

interface UpdateSurvey {
    id: number;
    title: string;
    participants: string[];
    creator: string;
    recipeVote: {[key: string]: number};
    options: Recipe[];
    creationDate: Date;
}

interface CreateSurvey {
    title: string;
    options: Recipe[];
}

export type {Survey, CreateSurvey,UpdateSurvey};

