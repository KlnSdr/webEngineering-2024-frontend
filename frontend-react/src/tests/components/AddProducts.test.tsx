import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import AddProducts from "../../components/AddProducts"; // Adjust the import if the path is different
import { NeededProduct, Product } from "../../types/Products"; // Adjust the import if the path is different
import exp from "constants";

jest.mock("../../components/AddProductLine", () => ({
                                                        initialValue,
                                                        onChange,
                                                        onRemove
                                                    }: any) => (
    <div>
        <input
            data-testid="product-input"
            defaultValue={initialValue}
            onChange={(e) => onChange(e.target.value, 1)}
        />
        <button data-testid="remove-button" onClick={onRemove}>Remove</button>
    </div>
));

describe("AddProducts Component", () => {
    const availableProducts: Product[] = [
        { name: "Product 1", unit: "kg", id: 0 },
        { name: "Product 2", unit: "ltr", id: 1 },
    ];

    const initialValue: NeededProduct[] = [
        { id: 1, productName: "Product 1", amount: 2 },
    ];

    test("renders with initial values", () => {
        const handleChange = jest.fn();

        const {asFragment} = render(<AddProducts initialValue={initialValue} onChange={handleChange} availableProducts={availableProducts} />);

        const productInput = screen.getByTestId("product-input");
        expect(productInput).toHaveValue(initialValue[0].productName);
        expect(asFragment()).toMatchSnapshot();
    });

    test("adds a new product line", () => {
        const handleChange = jest.fn();

        render(<AddProducts initialValue={initialValue} onChange={handleChange} availableProducts={availableProducts} />);

        const addButton = screen.getByTestId("add-product");
        fireEvent.click(addButton);

        const productInputs = screen.getAllByTestId("product-input");
        expect(productInputs).toHaveLength(initialValue.length + 1);
        expect(handleChange).toHaveBeenCalledWith(expect.arrayContaining([
            ...initialValue,
            expect.objectContaining({ productName: "Product 1", amount: 0 })
        ]));
    });

    test("updates a product's name and amount", () => {
        const handleChange = jest.fn();

        render(<AddProducts initialValue={initialValue} onChange={handleChange} availableProducts={availableProducts} />);

        const productInput = screen.getByTestId("product-input");
        fireEvent.change(productInput, { target: { value: "Product 2" } });

        expect(handleChange).toHaveBeenCalledWith(expect.arrayContaining([
            expect.objectContaining({ productName: "Product 2", amount: 1 })
        ]));
    });

    test("removes a product line", () => {
        const handleChange = jest.fn();

        render(<AddProducts initialValue={initialValue} onChange={handleChange} availableProducts={availableProducts} />);

        const removeButton = screen.getByTestId("remove-button");
        fireEvent.click(removeButton);

        expect(screen.queryByTestId("product-input")).not.toBeInTheDocument();
        expect(handleChange).toHaveBeenCalledWith([]);
    });
});
