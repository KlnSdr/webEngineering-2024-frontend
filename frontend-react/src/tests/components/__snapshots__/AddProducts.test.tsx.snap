// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProducts Component renders with initial values 1`] = `
<DocumentFragment>
  <div>
    <div>
      <input
        data-testid="product-input"
        value="Product 1"
      />
      <button
        data-testid="remove-button"
      >
        Remove
      </button>
    </div>
    <button
      class="bi bi-plus btn btn-primary"
      data-testid="add-product"
      type="button"
    />
  </div>
</DocumentFragment>
`;
